### PR TITLE
feat(ivy): add `ngcc` ivy switch

### PR DESCRIPTION
**Pull request by mhevery**
feat(ivy): add `ngcc` ivy switch

Provides a runtime and compile time switch for ivy including 
`ApplicationRef.bootstrapModule`.

This is done by naming the symbols such that `ngcc` (angular 
Compatibility compiler) can rename symbols in such a way that running
`ngcc` command will switch the `@angular/core` module from `legacy` to
`ivy` mode.

This is done as follows:

```
const someToken__PRE_NGCC__ = ‘legacy mode’;
const someToken__POST_NGCC__ = ‘ivy mode’;

export someSymbol = someToken__PRE_NGCC__;
```

The `ngcc` will search for any token which ends with `__PRE_NGCC__` 
and replace it with `__POST_NGCC__`. This allows the `@angular/core`
package to be rewritten to ivy mode post `ngcc` execution.

### CODE REVIEW

**Review by IgorMinar — COMMENTED** [review]
This seems to be heading in the right direction, but it appears incomplete and as I commented I think we can do this without ngcc rewriting stuff.

I left a bunch of notes. Please respond inline. thanks!

File: packages/core/src/ivy_switch_legacy.ts [hunk 1]
```diff
@@ -6,9 +6,60 @@
...
+const compileNgModuleQueue: any[] = [];
+const compilePipeQueue: any[] = [];
+
+const R3_COMPILE_COMPONENT__PRE_NGCC__: DirectiveCompiler = noop;
+const R3_COMPILE_DIRECTIVE__PRE_NGCC__: DirectiveCompiler = noop;
+const R3_COMPILE_INJECTABLE__PRE_NGCC__: DirectiveCompiler = preR3InjectableCompile;
+const R3_COMPILE_NGMODULE__PRE_NGCC__: DirectiveCompiler = preR3NgModuleCompile;
+const R3_COMPILE_PIPE__PRE_NGCC__: DirectiveCompiler = noop;
+
 export const ivyEnabled = false;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/ivy_switch_legacy.ts]: fixed

File: packages/core/src/ivy_switch_legacy.ts [hunk 2]
```diff
@@ -6,9 +6,60 @@
...
-export const R3_COMPILE_INJECTABLE: ((type: any, meta: any) => void)|null = null;
-export const R3_COMPILE_NGMODULE: ((type: any, meta: any) => void)|null = null;
-export const R3_COMPILE_PIPE: ((type: any, meta: any) => void)|null = null;
+export let R3_COMPILE_COMPONENT: DirectiveCompiler = R3_COMPILE_COMPONENT__PRE_NGCC__;
+export let R3_COMPILE_DIRECTIVE: DirectiveCompiler = R3_COMPILE_DIRECTIVE__PRE_NGCC__;
+export let R3_COMPILE_INJECTABLE: DirectiveCompiler = R3_COMPILE_INJECTABLE__PRE_NGCC__;
+export let R3_COMPILE_NGMODULE: DirectiveCompiler = R3_COMPILE_NGMODULE__PRE_NGCC__;
+export let R3_COMPILE_PIPE: DirectiveCompiler = R3_COMPILE_PIPE__PRE_NGCC__;
+
+export function enableIvyJitCompiler() {
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/ivy_switch_legacy.ts]: removed

File: packages/core/src/ivy_switch_legacy.ts [hunk 3]
```diff
@@ -6,9 +6,60 @@
...
+const R3_COMPILE_DIRECTIVE__POST_NGCC__ = compileDirective;
+const R3_COMPILE_INJECTABLE__POST_NGCC__ = compileInjectable;
+const R3_COMPILE_NGMODULE__POST_NGCC__ = compileNgModule;
+const R3_COMPILE_PIPE__POST_NGCC__ = compilePipe;
+
+const compileComponentQueue: any[] = [];
+const compileDirectiveQueue: any[] = [];
+const compileInjectableQueue: any[] = [];
+const compileNgModuleQueue: any[] = [];
+const compilePipeQueue: any[] = [];
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/ivy_switch_legacy.ts]: removed

File: packages/core/src/ivy_switch_legacy.ts [hunk 4]
```diff
@@ -6,9 +6,60 @@
...
+export let R3_COMPILE_INJECTABLE: DirectiveCompiler = R3_COMPILE_INJECTABLE__PRE_NGCC__;
+export let R3_COMPILE_NGMODULE: DirectiveCompiler = R3_COMPILE_NGMODULE__PRE_NGCC__;
+export let R3_COMPILE_PIPE: DirectiveCompiler = R3_COMPILE_PIPE__PRE_NGCC__;
+
+export function enableIvyJitCompiler() {
+  R3_COMPILE_COMPONENT = R3_COMPILE_COMPONENT__POST_NGCC__;
+  R3_COMPILE_DIRECTIVE = R3_COMPILE_DIRECTIVE__POST_NGCC__;
+  R3_COMPILE_INJECTABLE = R3_COMPILE_INJECTABLE__POST_NGCC__;
+  R3_COMPILE_NGMODULE = R3_COMPILE_NGMODULE__POST_NGCC__;
+  R3_COMPILE_PIPE = R3_COMPILE_PIPE__POST_NGCC__;
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/ivy_switch_legacy.ts]: discussed offline

File: packages/core/src/ivy_switch_legacy.ts [hunk 5]
```diff
@@ -6,9 +6,60 @@
...
+
+  processDecorators(compileComponentQueue, R3_COMPILE_COMPONENT__POST_NGCC__);
+  processDecorators(compileDirectiveQueue, R3_COMPILE_DIRECTIVE__POST_NGCC__);
+  processDecorators(compileInjectableQueue, R3_COMPILE_INJECTABLE__POST_NGCC__);
+  processDecorators(compileNgModuleQueue, R3_COMPILE_NGMODULE__POST_NGCC__);
+  processDecorators(compilePipeQueue, R3_COMPILE_PIPE__POST_NGCC__);
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/ivy_switch_legacy.ts]: reverted

File: packages/core/src/application_ref.ts [hunk 6]
```diff
@@ -9,28 +9,52 @@
...
+}
+
+function compileNgModuleFactory__POST_NGCC__<M>(
+    injector: Injector, options: CompilerOptions,
+    moduleType: Type<M>): Promise<NgModuleFactory<M>> {
+  ngDevMode && assertNgModuleType(moduleType);
+  return Promise.resolve(new R3NgModuleFactory(moduleType));
+}
+
+export function enableIvyBootstrap() {
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/application_ref.ts]: reverted

**Review by IgorMinar — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
LGTM but please note that CI is currently very unhappy.

**Comment by IgorMinar** [comment]
@mhevery can you please update the commit message? it still mentions the api that we decided not to expose.

**Comment by mhevery** [comment]
http://test/OCL:208886029:BASE:208903764:1534375685903:825ea3b6

**Pull request closed without merging**